### PR TITLE
Update free monad docs

### LINKED ISSUE #804: Update free monad tut for newer Free structure
The Cats `Free` monad has `Coyoneda` "built-in". This means that when using a custom ADT with `Free` and an interpreter, you don't need to create a `Functor` for your data type. This means that the ADT in our [doc](http://non.github.io/cats//tut/freemonad.html) could be simplified, which would probably be better to do for an introductory doc.

### PR DESCRIPTION
**Pull request by ceedubs**
This resolves #804.

I also updated the "pure" compiler to use `State`. It previously had
said that a pure compiler couldn't be written and used with `foldMap`,
but that's not really true if you use something like `State`. I think
this is probably a cleaner and more useful solution.

@mandubian @raulraja would you be available to review?

### CODE REVIEW

File: docs/src/main/tut/freemonad.md [hunk 1]
```diff
@@ -77,30 +77,21 @@ recursive values.
...
 
 There are six basic steps to "freeing" the ADT:
 
 1. Create a type based on `Free[_]` and `KVStoreA[_]`.
-2. Prove `KVStoreA[_]` has a functor.
-3. Create smart constructors for `KVStore[_]` using `liftF`.
-4. Build a program out of key-value DSL operations.
-5. Build a compiler for programs of DSL operations.
-6. Execute our compiled program.
+2. Create smart constructors for `KVStore[_]` using `liftF`.
```
Comment (raulraja) [inline docs/src/main/tut/freemonad.md]: An alternative to `liftF` is to use `Free.inject` when targeting `cats.data.Coproduct` as a composition mechanism maybe it should be mentioned here.

Comment (raulraja) [inline docs/src/main/tut/freemonad.md]: I guess it's mentioned below. Perhaps we should consider adding some kind of TOC at the beginning of each tutorial to show at a glance the different sections and topics being described.

**Comment by raulraja** [comment]
@ceedubs LGTM!. My comments are just suggestion for future consideration in case the tutorials become too long.

**Comment by non** [comment]
:+1:

**Comment by mandubian** [comment]
yes that was required! thanks for the effort ;)
:+1:

**Comment by ceedubs** [comment]
Thanks for the quick feedback everyone. @raulraja I agree that tables of contents might be helpful in docs. I'll create a separate issue to track that.

**Pull request merged**